core_v14/src/substrate_type.rs

```diff
@@ -1,275 +1,352 @@
-// Copyright 2019-2021 Parity Technologies (UK) Ltd.
-// This file is part of substrate-desub.
-//
-// substrate-desub is free software: you can redistribute it and/or modify
-// it under the terms of the GNU General Public License as published by
-// the Free Software Foundation, either version 3 of the License, or
-// (at your option) any later version.
-//
-// substrate-desub is distributed in the hope that it will be useful,
-// but WITHOUT ANY WARRANTY; without even the implied warranty of
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-// GNU General Public License for more details.
-//
-// You should have received a copy of the GNU General Public License
-// along with substrate-desub.  If not, see <http://www.gnu.org/licenses/>.
-use serde::{Deserialize, Serialize};
-use std::fmt::{self, Display};
+use scale_info::PortableRegistry;
+use std::fmt::{ Debug };
+use crate::util::{ for_each_between, ForEachBetween };
 
-/// A field with an associated name
-#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
-pub struct StructField {
-	pub name: String,
-	pub ty: SubstrateType,
-}
+/// The portable form of a scale info type definition, as obtained from metadata.
+type ScaleInfoTypeDef = scale_info::TypeDef<scale_info::form::PortableForm>;
 
-impl Display for StructField {
-	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-		write!(f, "struct_field({}: {})", self.name, self.ty)
-	}
-}
+/// The portable form of a [`scale_info::Type`], as obtained from metadata.
+type ScaleInfoType = scale_info::Type<scale_info::form::PortableForm>;
 
-impl StructField {
-	pub fn new<S: Into<String>>(name: S, ty: SubstrateType) -> Self {
-		let name = name.into();
-		Self { name, ty }
-	}
+/// This roughly mirrors [`scale_info::TypeDef`], but is slightly simplified
+/// and fully public, so that it's possible to construct types ourselves.
+pub enum SubstrateType {
+    Composite(CompositeType),
+    Variant(VariantType),
+    Sequence(SequenceType),
+    Array(ArrayType),
+    Tuple(TupleType),
+    Primitive(PrimitiveType),
+    Compact(CompactType),
+    BitSequence(BitSequenceType),
```

Comment:
Isn't this one feature gated in `scale-info`?

Reply:
It is! Honestly I've no idea whether or not it's necessary in the decoding, but the <=V13 logic handled decoding to bitvec, so I did too. I did more recently run into some sporadic errors that meant I had to add a feature flag, though...